src/uct/base/uct_iov.inl

```diff
@@ -0,0 +1,123 @@
+ *
+ * @param [in] uct_iov_iter    Pointer to the UCT IOV iterator.
+ */
+static UCS_F_ALWAYS_INLINE
+void uct_iov_iter_init(uct_iov_iter_t *uct_iov_iter)
+{
+    uct_iov_iter->iov_offset    = 0;
+    uct_iov_iter->buffer_offset = 0;
+}
+
+/**
+ * Fill IOVEC data structure by data provided in the array of UCT IOVs.
+ * The function avoids copying IOVs with zero length.
+ *
+ * @param [out]    io_vec          Pointer to the resulted array of IOVECs (must
+ *                                 have at least @uct_iov_cnt elements).
+ * @param [in/out] io_vec_cnt_p    Pointer to the varibale that holds the number
+ *                                 of the elements in the array of IOVECs (input:
+ *                                 initialial, out: result).
+ * @param [in]     uct_iov         Pointer to the array of UCT IOVs.
+ * @paran [in]     uct_iov_cnt     Number of the elemnts in the array of UCT IOVs.
+ * @param [in]     max_length      The maximal total length of bytes that should be
+ *                                 in the resulted array of IOVECs.
+ * @param [in]     uct_iov_iter    Pointer to the UCT IOV iterator.
+ *
+ * @return The total length of the resulted array of IOVECs.
+ */
+static UCS_F_ALWAYS_INLINE
+size_t uct_iov_to_io_vec(struct iovec *io_vec, size_t *io_vec_cnt_p,
+                         const uct_iov_t *uct_iov, size_t uct_iov_cnt,
+                         size_t max_length, uct_iov_iter_t *uct_iov_iter)
+{
+    size_t total_length = 0;
+    size_t io_vec_it    = 0;
+    size_t uct_iov_it, io_vec_length;
+
+    for (uct_iov_it = uct_iov_iter->iov_offset; (uct_iov_it < uct_iov_cnt) &&
+                                                (max_length != 0) &&
+                                                (io_vec_it < *io_vec_cnt_p);) {
+        ucs_assert(uct_iov_get_length(&uct_iov[uct_iov_it]) >=
+                   uct_iov_iter->buffer_offset);
+        io_vec_length = uct_iov_get_length(&uct_iov[uct_iov_it]) -
+                        uct_iov_iter->buffer_offset;
+        if (io_vec_length == 0) {
+            /* Avoid zero length elements in resulted iov_vec */
+            ++uct_iov_it;
+            continue;
+        }
+
+        io_vec[io_vec_it].iov_len   = ucs_min(io_vec_length, max_length);
+        io_vec[io_vec_it].iov_base  = UCS_PTR_BYTE_OFFSET(uct_iov[uct_iov_it].buffer,
+                                                          uct_iov_iter->buffer_offset);
+
+       if (io_vec_length > max_length) {
+            uct_iov_iter->buffer_offset += max_length;
+        } else {
+            ucs_assert((uct_iov_iter->buffer_offset == 0) ||
+                       (uct_iov_it == uct_iov_iter->iov_offset));
+            uct_iov_iter->buffer_offset = 0;
+            ++uct_iov_it;
+        }
+
+        ucs_assert(max_length >= io_vec[io_vec_it].iov_len);
+        total_length          += io_vec[io_vec_it].iov_len;
+        max_length            -= io_vec[io_vec_it].iov_len;
```

Comment:
typically we try to avoid reusing stack vars. @yosefe is it code-style guide line?

Reply:
yes, it can be better to define `remain_length`and use it in the code
we could add some useful `ucs_assert()`s